libwnbd/wnbd_ioctl.cpp

```diff
@@ -133,6 +137,315 @@ DWORD WnbdOpenAdapterCMDeviceInstance(PDEVINST DeviceInstance)
+    if (INVALID_HANDLE_VALUE == *AdapterDevHandle) {
+        Status = GetLastError();
+        LogError(
+            "SetupDiGetClassDevs failed. "
+            "Error: %d. Error message: %s", Status, win32_strerror(Status).c_str());
+        return Status;
+    }
+
+    for (DWORD I = 0; !Found && SetupDiEnumDeviceInfo(*AdapterDevHandle, I, DevInfoData); I++)
+    {
+        if (!SetupDiGetDeviceRegistryPropertyA(*AdapterDevHandle, DevInfoData, SPDRP_HARDWAREID, 0,
+            (PBYTE)TempBuf, sizeof(TempBuf) - 1, 0)) {
+            continue;
+        }
+
+        if (strstr(TempBuf, WNBD_HARDWAREID) != NULL) {
+            Found = TRUE;
+            break;
+        }
+    }
+
+    if (!Found) {
+        Status = GetLastError();
+        LogInfo("Error: %d. Error message: %s", Status, win32_strerror(Status).c_str());
+        return Status;
+    }
+
+    return ERROR_SUCCESS;
+}
+
+DWORD HardRemoveAllDevices()
+{
+    DWORD BufferSize = 0;
+    DWORD Status = 0;
+    PWNBD_CONNECTION_LIST TempList = NULL;
+
+    WNBD_OPTION_VALUE OptValue = { WnbdOptBool };
+    /* Disallow new mappings so we can remove all current mappings */
+    OptValue.Data.AsBool = FALSE;
+    Status = WnbdSetDrvOpt("NewMappingsAllowed", &OptValue, FALSE);
+    if (Status) {
+        LogError(
+            "WnbdSetDrvOpt Error: %d. Error message: %s", Status, win32_strerror(Status).c_str());
+        goto exit;
+    }
+
+    Status = WnbdList(TempList, &BufferSize);
+    if (!BufferSize) {
+        goto exit;
+    }
+    TempList = (PWNBD_CONNECTION_LIST)calloc(1, BufferSize);
+    if (!TempList) {
+        Status = ERROR_NOT_ENOUGH_MEMORY;
+        LogError(
+            "calloc Error: %d. Error message: %s", Status, win32_strerror(Status).c_str());
+        goto exit;
+    }
+    Status = WnbdList(TempList, &BufferSize);
+    if (Status) {
+        LogError(
+            "WnbdList Error: %d. Error message: %s", Status, win32_strerror(Status).c_str());
+        goto exit;
+    }
+    CHAR* InstanceName;
+    for (ULONG index = 0; index < TempList->Count; index++) {
+        InstanceName = TempList->Connections[index].Properties.InstanceName;
+        WNBD_REMOVE_OPTIONS RemoveOptions = { 0 };
+        RemoveOptions.Flags.HardRemove = TRUE;
+        Status = WnbdRemoveEx(InstanceName, &RemoveOptions);
+        if (Status) {
+            LogError(
+                "Could not remove device: %s. "
+                "Error: %d. Error message: %s",
+                InstanceName, Status, win32_strerror(Status).c_str());
+            goto exit;
+        }
+    }
+
+exit:
+    if (TempList) {
+        free(TempList);
+    }
+    return Status;
+}
+
+DWORD WnbdUninstallDriver(PBOOL RebootRequired)
+{
+    DWORD Status;
+    HDEVINFO AdapterDevHandle = INVALID_HANDLE_VALUE;
+    SP_DEVINFO_DATA DevInfoData;
+    DevInfoData.cbSize = sizeof DevInfoData;
+    Status = FindDevice(&AdapterDevHandle, &DevInfoData);
```

Comment:
Just because there isn't a WNBD device doesn't mean that the driver is not installed. We should go ahead and actually remove the driver.